File: frontend/test/lib/rbac.test.ts
import { afterEach, describe, expect, it, vi } from 'vitest';
import {
  NO_ASK,
  askable,
  fetchRBAC,
  fetchWho,
  grantLabel,
  grantNote,
  matches,
  ruleLabel,
  whereLabel,
  whoFailure,
} from '../../src/lib/rbac';
import type { RBACGrant, RBACSubject } from '../../src/lib/types';

function stub(body: unknown, ok = true, status = 200) {
  const fetcher = vi.fn((url: string) => {
    void url;
    return Promise.resolve({ ok, status, json: () => Promise.resolve(body) });
  });
  vi.stubGlobal('fetch', fetcher);
  return fetcher;
}

function subject(extra: Partial<RBACSubject> = {}): RBACSubject {
  return { kind: 'User', name: 'ana', label: 'ana', grants: [], ...extra };
}

function grant(extra: Partial<RBACGrant> = {}): RBACGrant {
  return {
    binding: 'read',
    bindingKind: 'ClusterRoleBinding',
    role: 'reader',
    roleKind: 'ClusterRole',
    ...extra,
  };
}

afterEach(() => {
  vi.unstubAllGlobals();
});

describe('reading the index', () => {
  it('reads the subjects back', async () => {
    stub({ subjects: [subject()] });

    expect((await fetchRBAC()).subjects).toHaveLength(1);
  });

  it('fills in an empty answer', async () => {
    stub({});

    expect((await fetchRBAC()).subjects).toEqual([]);
  });

  it('reports a failure with its status', async () => {
    stub({ message: 'no' }, false, 500);

    await expect(fetchRBAC()).rejects.toThrow();
  });
});

describe('asking who can', () => {
  it('needs a verb and a resource', () => {
    expect(askable(NO_ASK)).toBe(false);
    expect(askable({ ...NO_ASK, verb: 'create' })).toBe(false);
    expect(askable({ ...NO_ASK, verb: 'create', resource: 'pods' })).toBe(true);
  });

  it('treats blank space as nothing asked', () => {
    expect(askable({ ...NO_ASK, verb: '  ', resource: ' ' })).toBe(false);
  });

  it('carries the verb and resource', async () => {
    const fetcher = stub({ subjects: [] });

    await fetchWho({ ...NO_ASK, verb: 'create', resource: 'pods/exec' });

    expect(fetcher.mock.calls[0][0]).toContain('verb=create');
    expect(fetcher.mock.calls[0][0]).toContain('pods%2Fexec');
  });

  it('leaves out the group and namespace when they are blank', async () => {
    const fetcher = stub({ subjects: [] });

    await fetchWho({ ...NO_ASK, verb: 'get', resource: 'secrets' });

    expect(fetcher.mock.calls[0][0]).not.toContain('group=');
    expect(fetcher.mock.calls[0][0]).not.toContain('namespace=');
  });

  it('carries the group and namespace when they are given', async () => {
    const fetcher = stub({ subjects: [] });

    await fetchWho({ verb: 'create', resource: 'deployments', group: 'apps', namespace: 'web' });

    expect(fetcher.mock.calls[0][0]).toContain('group=apps');
    expect(fetcher.mock.calls[0][0]).toContain('namespace=web');
  });

  it('reports a failure', async () => {
    stub({}, false, 400);

    await expect(fetchWho({ ...NO_ASK, verb: 'get', resource: 'pods' })).rejects.toThrow();
  });
});

describe('when the question itself fails', () => {
  it('says what failed and why', async () => {
    stub({ message: 'no' }, false, 400);

    await expect(fetchWho({ ...NO_ASK, verb: 'get', resource: 'pods' })).rejects.toThrow();
  });
});

describe('what a row says', () => {
  it('says everywhere for a subject with no namespaces', () => {
    expect(whereLabel(subject())).toBe('everywhere');
  });

  it('lists a few namespaces', () => {
    expect(whereLabel(subject({ namespaces: ['db', 'web'] }))).toBe('db, web');
  });

  it('counts the rest when there are many', () => {
    expect(whereLabel(subject({ namespaces: ['a', 'b', 'c', 'd', 'e'] }))).toBe(
      'a, b, c and 2 more',
    );
  });

  it('names the binding that made a grant', () => {
    expect(grantLabel(grant())).toBe('ClusterRoleBinding read → ClusterRole reader · everywhere');
  });

  it('names the namespace a grant is confined to', () => {
    expect(grantLabel(grant({ namespace: 'web' }))).toContain('· web');
  });

  it('says when a grant names a role that is not there', () => {
    expect(grantNote(grant({ missing: true }))).toBe('the role it names does not exist');
  });

  it('says when an aggregated role has not been filled in', () => {
    expect(grantNote(grant({ aggregated: true, rules: [] }))).toContain('controller');
  });

  it('says nothing about an ordinary grant', () => {
    expect(grantNote(grant({ rules: [{ verbs: ['get'] }] }))).toBe('');
  });

  it('reads a rule as verbs on resources', () => {
    expect(ruleLabel({ verbs: ['get', 'list'], resources: ['pods'], groups: [''] })).toBe(
      'get, list on pods',
    );
  });

  it('names the api group when there is one', () => {
    expect(ruleLabel({ verbs: ['get'], resources: ['deployments'], groups: ['apps'] })).toBe(
      'get on deployments in apps',
    );
  });

  it('says when a rule is limited to named objects', () => {
    expect(ruleLabel({ verbs: ['get'], resources: ['secrets'], names: ['token'] })).toContain(
      'named token',
    );
  });

  it('says nothing rather than blank for an empty rule', () => {
    expect(ruleLabel({})).toBe('nothing on nothing');
  });
});

describe('when the question fails', () => {
  it('says what failed and why', () => {
    expect(whoFailure(new Error('the apiserver refused'))).toBe(
      'Asking who can: the apiserver refused',
    );
  });

  it('says what failed when there is no why', () => {
    expect(whoFailure('nope')).toBe('Asking who can failed');
  });
});

describe('filtering the list', () => {
  it('keeps everything for an empty query', () => {
    expect(matches(subject(), '  ')).toBe(true);
  });

  it('matches on the subject name', () => {
    expect(matches(subject({ label: 'system:serviceaccount:web:api' }), 'web')).toBe(true);
  });

  it('matches on a power', () => {
    expect(matches(subject({ powers: ['reads secrets'] }), 'secret')).toBe(true);
  });

  it('drops what does not match', () => {
    expect(matches(subject(), 'zzz')).toBe(false);
  });
});

File: internal/server/rbac_test.go
package server

import (
	"context"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/sophotechlabs/spinoza/internal/api"
	"github.com/sophotechlabs/spinoza/internal/rbac"
)

type indexed struct {
	notStubbed

	index rbac.Index
}

func (i *indexed) RBACIndex(context.Context) rbac.Index {
	return i.index
}

func holding(powers []string, grants ...rbac.Grant) rbac.Holder {
	return rbac.Holder{
		Subject: rbac.Subject{Kind: rbac.KindUser, Name: "ana"},
		Powers:  powers,
		Grants:  grants,
	}
}

func granted(role, namespace string, rules ...rbac.Rule) rbac.Grant {
	return rbac.Grant{
		Binding:     role + "-binding",
		BindingKind: "ClusterRoleBinding",
		Role:        role,
		RoleKind:    rbac.ClusterRoleKind,
		Namespace:   namespace,
		Rules:       rules,
	}
}

func canDo(verb, resource string) rbac.Rule {
	return rbac.Rule{Verbs: []string{verb}, Groups: []string{""}, Resources: []string{resource}}
}

func rbacServer(t *testing.T, index rbac.Index) *httptest.Server {
	t.Helper()
	return stubbedServer(t, &indexed{index: index})
}

func TestTheIndexNamesEverySubjectAndWhatItHolds(t *testing.T) {
	ts := rbacServer(t, rbac.Index{
		Holders: []rbac.Holder{holding([]string{"reads secrets"}, granted("reader", "", canDo("get", "secrets")))},
	})

	var got api.RBACIndex
	readFleet(t, ts, "/api/rbac", &got)

	if len(got.Subjects) != 1 {
		t.Fatalf("subjects = %+v", got.Subjects)
	}
	one := got.Subjects[0]
	if one.Label != "ana" || len(one.Powers) != 1 {
		t.Fatalf("the subject came back as %+v", one)
	}
	if len(one.Grants) != 1 || one.Grants[0].Role != "reader" {
		t.Fatalf("grants = %+v", one.Grants)
	}
}

func TestASubjectSaysWhereItsGrantsApply(t *testing.T) {
	ts := rbacServer(t, rbac.Index{
		Holders: []rbac.Holder{holding(nil, granted("reader", "web", canDo("get", "pods")))},
	})

	var got api.RBACIndex
	readFleet(t, ts, "/api/rbac", &got)

	if len(got.Subjects[0].Namespaces) != 1 || got.Subjects[0].Namespaces[0] != "web" {
		t.Fatalf("namespaces = %+v", got.Subjects[0].Namespaces)
	}
}

func TestTheIndexCarriesTheRulesAsWritten(t *testing.T) {
	ts := rbacServer(t, rbac.Index{
		Holders: []rbac.Holder{holding(nil, granted("wide", "",
			rbac.Rule{Verbs: []string{"*"}, Groups: []string{"*"}, Resources: []string{"*"}}))},
	})

	var got api.RBACIndex
	readFleet(t, ts, "/api/rbac", &got)

	rule := got.Subjects[0].Grants[0].Rules[0]
	if len(rule.Verbs) != 1 || rule.Verbs[0] != "*" {
		t.Fatalf("the rule came back as %+v", rule)
	}
}

func TestTheIndexSaysWhatItCouldNotSee(t *testing.T) {
	ts := rbacServer(t, rbac.Index{
		Absent: []string{"RoleBinding web/read wants Role gone"},
		Error:  "not discovered: clusterroles",
	})

	var got api.RBACIndex
	readFleet(t, ts, "/api/rbac", &got)

	if len(got.Absent) != 1 || got.Error != "not discovered: clusterroles" {
		t.Fatalf("index = %+v", got)
	}
}

func TestAnEmptyIndexComesBackAsAList(t *testing.T) {
	ts := rbacServer(t, rbac.Index{})

	var got api.RBACIndex
	readFleet(t, ts, "/api/rbac", &got)

	if got.Subjects == nil {
		t.Fatal("subjects came back as null rather than an empty list")
	}
}

func TestTheIndexStopsAtItsCapAndSaysSo(t *testing.T) {
	held := rbac.Index{}
	for range rbacSubjectCap + 20 {
		held.Holders = append(held.Holders, holding(nil, granted("reader", "web", canDo("get", "pods"))))
	}
	ts := rbacServer(t, held)

	var got api.RBACIndex
	readFleet(t, ts, "/api/rbac", &got)

	if len(got.Subjects) != rbacSubjectCap {
		t.Fatalf("subjects = %d", len(got.Subjects))
	}
	if got.Dropped != 20 {
		t.Fatalf("dropped = %d", got.Dropped)
	}
}

func TestWhoNamesTheSubjectsThatMay(t *testing.T) {
	ts := rbacServer(t, rbac.Index{
		Holders: []rbac.Holder{
			holding(nil, granted("exec", "", canDo("create", "pods/exec"))),
			{
				Subject: rbac.Subject{Kind: rbac.KindUser, Name: "bo"},
				Grants:  []rbac.Grant{granted("look", "", canDo("get", "pods"))},
			},
		},
	})

	var got api.RBACIndex
	readFleet(t, ts, "/api/rbac/who?verb=create&resource=pods/exec", &got)

	if len(got.Subjects) != 1 || got.Subjects[0].Label != "ana" {
		t.Fatalf("who = %+v", got.Subjects)
	}
}

func TestWhoNeedsAVerbAndAResource(t *testing.T) {
	ts := rbacServer(t, rbac.Index{})

	for _, query := range []string{"", "?verb=get", "?resource=pods"} {
		resp, _ := doRequest(t, http.MethodGet, ts.URL+"/api/rbac/who"+query, nil)
		if resp.StatusCode != http.StatusBadRequest {
			t.Fatalf("status for %q = %d", query, resp.StatusCode)
		}
	}
}

func TestWhoNarrowsToANamespace(t *testing.T) {
	ts := rbacServer(t, rbac.Index{
		Holders: []rbac.Holder{holding(nil, granted("reader", "web", canDo("get", "secrets")))},
	})

	var got api.RBACIndex
	readFleet(t, ts, "/api/rbac/who?verb=get&resource=secrets&namespace=other", &got)

	if len(got.Subjects) != 0 {
		t.Fatalf("who = %+v, want nobody outside the namespace", got.Subjects)
	}
}

func TestWhoCarriesTheApiGroup(t *testing.T) {
	ts := rbacServer(t, rbac.Index{
		Holders: []rbac.Holder{holding(nil, granted("deployer", "",
			rbac.Rule{Verbs: []string{"create"}, Groups: []string{"apps"}, Resources: []string{"deployments"}}))},
	})

	var got api.RBACIndex
	readFleet(t, ts, "/api/rbac/who?verb=create&group=apps&resource=deployments", &got)

	if len(got.Subjects) != 1 {
		t.Fatalf("who = %+v", got.Subjects)
	}
}
